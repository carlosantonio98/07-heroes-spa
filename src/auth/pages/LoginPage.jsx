import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../context/AuthContext';

export const LoginPage = () => {
    
    const { login } = useContext( AuthContext );
    const navigate = useNavigate();

    const onLogin = () => {

        const lastPath = localStorage.getItem('lastPath') || '/';  // leemos el lastPath del localStorage para redirigirnos a la ultima ruta a la que accedimos cuando estabamos loggeado.

        login( 'Carlos Camacho' );

        navigate( lastPath, {
            replace: true  // colocamos el replace: true para que ya no regresemos al login si ya lo pasamos, le decimos que borre el historial
        });
    }

    return (
        <div className="container mt-5">
            <h1>LoginPage</h1>
            <hr />
            <button
                className="btn btn-primary"
                onClick={ onLogin }
            >
                Login
            </button>
        </div>
    );
}
