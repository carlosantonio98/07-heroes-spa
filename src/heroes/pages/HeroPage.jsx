import { useMemo } from 'react';
import { Navigate, useNavigate, useParams } from 'react-router-dom';
import { getHeroesById } from '../helpers/gerHeroById';

export const HeroPage = () => {
    const navigate = useNavigate();
    const { id } = useParams();  // customHook creado por react-router-dom y nos sirve para acceder a los parametros de nuestra url que se almacenan en el Route.provider
    // const hero = getHeroesById( id );  // Es sumamente importate sacar esta variable fuera del funtional component dado el caso en el que si nuestro componente cambiara su estado y este se tuviera que tener que redibujar todos esto de nuevo. La variable se debe de sacar de aquí para no ejecutar la  función cada que este tenga un pequeño cambio, en este proyecto no lo sacaremos debido a que este no modifícara su estado.
    const hero = useMemo( ()=> getHeroesById( id ), [ id ])  // Memorizamos el valor del hero, no cambiara por nada en el mundo a menos de que el id de sus dependencias cambie, menorizamos para que no se redibuje todo el componente y no se vuelva a ejecutar esta función, solo cambiara si el id es modificado

    const onNavigateBack = () => {
        return navigate(-1);  // esto te lleva a la ruta anterior, esto no significa que no podamos hacer una condición mas elaborada
    }




    // Como recomendación debemos de colocar este tipo de condición que retorna un jsx arriba del return padre, esto para seguir una logica en la programación
    if ( !hero ) {
        return <Navigate to="/marvel" />;  // Navigate es un jsx que podemos regresar en lugar de un <>404 - Not Found</>
    }

    return (
        <div className="row mt-5">
            <div className="col-4">
                <img 
                    src={ `/assets/heroes/${ id }.jpg` }
                    alt={ hero.superhero } 
                    className="img-thumbnail animate__animated animate__fadeInLeft"
                />
            </div>

            <div className="col-8">
                <h3>{ hero.superhero }</h3>
                <ul className='list-group list-group-flush'>
                    <li className='list-group-item'> <b> Alter ego: </b> { hero.alter_ego } </li>
                    <li className='list-group-item'> <b> Publisher: </b> { hero.publisher } </li>
                    <li className='list-group-item'> <b> First appearance: </b> { hero.first_appearance } </li>
                </ul>

                <h5 className='mt-3'>Characters</h5>
                <p>{ hero.characters }</p>

                <button 
                    className='btn btn-outline-primary'
                    onClick={ onNavigateBack }
                >
                    Back
                </button>
            </div>
        </div>
    );
}