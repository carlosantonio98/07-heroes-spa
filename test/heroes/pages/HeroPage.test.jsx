import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { HeroPage } from '../../../src/heroes/pages/HeroPage';
import { MarvelPage } from '../../../src/heroes/pages/MarvelPage';


const mockedUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}));

describe('Pruebas en <HeroPage />', () => {

    const heroId = 'marvel-spider';

    beforeEach(() => jest.clearAllMocks());

    test('Debe de mostrarse correctamente con valores por defecto', () => {

        const { container } = render(
            <MemoryRouter initialEntries={['/hero/' + heroId]}>
                <Routes>
                    <Route path='/hero/:id' element={ <HeroPage /> } />
                </Routes>
            </MemoryRouter>    
        );

        expect( container ).toMatchSnapshot();

    });

    test('Debe de mostrar un heroe', () => {

        render(
            <MemoryRouter initialEntries={['/hero/' + heroId]}>
                <Routes>
                    <Route path='/hero/:id' element={ <HeroPage /> } />  {/* Para que detecte el params se debe de envolver el elemento en un route */}
                </Routes>
            </MemoryRouter>    
        );
        
        expect( screen.getByText('Spider Man') ).toBeTruthy();
        expect( screen.getByText('Alter ego:') ).toBeTruthy();
        expect( screen.getByText('Publisher:') ).toBeTruthy();
        expect( screen.getByText('First appearance:') ).toBeTruthy();
        expect( screen.getByText('Characters') ).toBeTruthy();

    });
    
    test('Debe de navegar a marvel si el heroe no existe', () => {

        render(
            <MemoryRouter initialEntries={['/hero/' + heroId + '-test']}>
                <Routes>
                    <Route path='/hero/:id' element={ <HeroPage /> } />
                    <Route path='/marvel' element={ <MarvelPage /> } />
                </Routes>
            </MemoryRouter>    
        );

        expect( screen.getByText('MarvelPage') ).toBeTruthy();

    });

    test('Debe de navegar para atrÃ¡s', () => {

        render(
            <MemoryRouter initialEntries={['/hero/' + heroId]}>
                <Routes>
                    <Route path='/hero/:id' element={ <HeroPage /> } />
                </Routes>
            </MemoryRouter>
        );

        const btn = screen.getByRole('button');
        fireEvent.click(btn);

        expect( mockedUseNavigate ).toHaveBeenCalled();

    });

});