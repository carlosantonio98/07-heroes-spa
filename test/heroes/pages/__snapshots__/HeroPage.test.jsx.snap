// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroPage /> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    class="row mt-5"
  >
    <div
      class="col-4"
    >
      <img
        alt="Spider Man"
        class="img-thumbnail animate__animated animate__fadeInLeft"
        src="/assets/heroes/marvel-spider.jpg"
      />
    </div>
    <div
      class="col-8"
    >
      <h3>
        Spider Man
      </h3>
      <ul
        class="list-group list-group-flush"
      >
        <li
          class="list-group-item"
        >
           
          <b>
             Alter ego: 
          </b>
           
          Peter Parker
           
        </li>
        <li
          class="list-group-item"
        >
           
          <b>
             Publisher: 
          </b>
           
          Marvel Comics
           
        </li>
        <li
          class="list-group-item"
        >
           
          <b>
             First appearance: 
          </b>
           
          Amazing Fantasy #15
           
        </li>
      </ul>
      <h5
        class="mt-3"
      >
        Characters
      </h5>
      <p>
        Peter Parker
      </p>
      <button
        class="btn btn-outline-primary"
      >
        Back
      </button>
    </div>
  </div>
</div>
`;
